template/packages/auth/README.md

```diff
@@ -0,0 +1,93 @@
+## the package tentatively titled "@Originate/auth"
+
+This package builds on top of `@Originate/leash` to provide endpoints that implement a common user-password authentication scheme on user-facing websites. The endpoints are documented below.
+
+### POST `/api/auth/login`
+
+``` tsx
+export interface LoginRequest {
+  id: string;
+  password: string;
+}
+
+export interface LoginGood<TUser> {
+  ok: true;
+  id: string;
+  user: TUser;
+  token: string;
+}
+
+export type LoginBad = {ok: false; key: string; error: string};
+
+export type LoginResponse<TUser> = LoginGood<TUser> | LoginBad;
+```
+
+Returns a stateless JWT token on success. This token, which will be a short alphanumeric string, authenticates further HTTP requests to the backend. Send it via an `Authorization: Bearer <token>` header. Tokens expire after a certain interval, which can be configured when the endpoints are installed (in the `@Originate/leash` sense of the word "install"). Tokens are stateless, meaning they are not stored in the backend. They are, instead, generated from an authenticated hash of the user's secret password, keyed off a key provided to the backend via its environment variables.
+
+### POST `/api/auth/signup`
+
+``` tsx
+export interface SignupRequest<TUser> {
+  user: TUser;
+  password: string;
+}
+
+export interface SignupGood {
+  ok: true;
+  id: string;
+}
+
+export type SignupBad = {ok: false; key: string; error: string};
+
+export type SignupResponse = SignupGood | SignupBad;
+```
+
+Creates a user account. The consumer of this library can decide what type the user object should be via a generic parameter `TUser`. For example, one app might want (full name, address, social-security number). Another app might want (Twitch username, Twitter username, Battle.net username). Along with providing a custom type, this library's consumers should also implement an `io-ts` Decoder for the type, so that the library can validate and parse incoming user objects.
+
+It is required to send an `Authorization: Bearer signup` header along with this request. We treat "signup" as a temporary token that authenticates an anonymous request (users who are signing up have yet to authenticate with the backend, because they are unable to yet, so they are effectively anonymous) and authorizes the request to access exactly one endpoint, which is this one, the signup endpoint. This requirement prevents trivial CSRF attacks.
```

Comment:
I wonder if this is necessary.  Can the endpoint work without and Authorization header?

Reply:
@mjhm without this, an attacker could simulate POSTs to this endpoint from a cross-origin context. the other way would be to mandate a csrf token, but then we would need the split the routes into mobile and web variants (as well as introducing the rather annoying csrf machinery on the frontend)

Reply:
Hmm -- I'm trying to imagine the actual attack vector here.  My understanding of CSRF counter measures are for situations where an attacker is trying to perform some action impersonating someone who is already authenticated (for example in another browser tab).
It seems to me that the worst thing that can happen without the "Auth.: Bearer signup" is that the user could be signed up to the service from a form on a foreign domain.  But the Auth token doesn't really prevent that possibility anyway.
These things are pretty brain-bending to me so I could very well be missing something.

Reply:
You can imagine a simple CSRF attack like this:

```html
<html>
  <body>
  <script>history.pushState('', '', '/')</script>
    <form action="https://www.example.com/" method="POST" enctype="text/plain">
      <input type="hidden" name='{"name":"value"}//' value="" />
    </form>
    <script>
      document.forms[0].submit();
    </script>
  </body>
</html>
```

as [described in this SE post](https://security.stackexchange.com/questions/170477/csrf-with-json-post-when-content-type-must-be-application-json).

There's a fuller description of this attack [here](https://owasp.org/www-community/attacks/csrf). The attack vector would be any cross-origin context submitting a POST.

The general principle is that any endpoint that changes the server state should be rendered inoperable when coming from a cross-origin context, I think. The worst thing that could happen without CRSF mitigation here is that someone could sign up a user, which isn't awful, but it could be combined with other attacks (like a DOS attack) to produce something worse. The idea behind having an Authorization header is to implement a form of defense in depth, if that makes sense?

Reply:
Ok, now you're making me hit the books.  Lots of interesting stuff in this area that would be worth reviewing at some point.  For example not exactly relevant to this discussion, but a related attack that I didn't know about is the Login CSRF attack.  See section 3 of https://seclab.stanford.edu/websec/csrf/csrf.pdf